flatten.go

```diff
@@ -0,0 +1,159 @@
+			return flattenIntersection(typed.IntermediateNode.ChildNodes)
+		case api.SetOperationUserset_EXCLUSION:
+			return flattenExclusion(typed.IntermediateNode.ChildNodes)
+		}
+	case *api.RelationTupleTreeNode_LeafNode:
+		var toReturn userSet = make(map[string]struct{})
+		for _, usr := range typed.LeafNode.Users {
+			toReturn.Add(usr)
+		}
+		return toReturn.ToSlice()
+	}
+	return nil
+}
+
+func flattenUnion(children []*api.RelationTupleTreeNode) []*api.User {
+	var toReturn userSet = make(map[string]struct{})
+	for _, child := range children {
+		toReturn.Add(flatten(child)...)
+	}
+	return toReturn.ToSlice()
+}
+
+func flattenIntersection(children []*api.RelationTupleTreeNode) []*api.User {
+	firstChildChildren := flatten(children[0])
+
+	if len(children) == 1 {
+		return firstChildChildren
+	}
+
+	var inOthers userSet = make(map[string]struct{})
+	inOthers.Add(flattenIntersection(children[1:])...)
+
+	maxChildren := len(firstChildChildren)
+	if len(inOthers) < maxChildren {
+		maxChildren = len(inOthers)
+	}
+
+	toReturn := make([]*api.User, 0, maxChildren)
+	for _, child := range firstChildChildren {
+		if inOthers.Contains(child) {
+			toReturn = append(toReturn, child)
+		}
+	}
+
+	return toReturn
+}
+
+func flattenExclusion(children []*api.RelationTupleTreeNode) []*api.User {
+	firstChildChildren := flatten(children[0])
+
+	if len(children) == 1 || len(firstChildChildren) == 0 {
+		return firstChildChildren
+	}
+
+	var toReturn userSet = make(map[string]struct{})
+	toReturn.Add(firstChildChildren...)
+	for _, child := range children[1:] {
+		toReturn.Remove(flatten(child)...)
+	}
+
+	return toReturn.ToSlice()
+}
+
+func leaf(start *api.ObjectAndRelation, children ...*api.User) *api.RelationTupleTreeNode {
+	return &api.RelationTupleTreeNode{
+		NodeType: &api.RelationTupleTreeNode_LeafNode{
+			LeafNode: &api.DirectUserset{
+				Users: children,
+			},
+		},
+		Expanded: start,
+	}
+}
+
+func setResult(
+	op api.SetOperationUserset_Operation,
+	start *api.ObjectAndRelation,
+	children []*api.RelationTupleTreeNode,
+) *api.RelationTupleTreeNode {
+	return &api.RelationTupleTreeNode{
+		NodeType: &api.RelationTupleTreeNode_IntermediateNode{
+			IntermediateNode: &api.SetOperationUserset{
+				Operation:  op,
+				ChildNodes: children,
+			},
+		},
+		Expanded: start,
+	}
+}
+
+func union(start *api.ObjectAndRelation, children ...*api.RelationTupleTreeNode) *api.RelationTupleTreeNode {
+	return setResult(api.SetOperationUserset_UNION, start, children)
+}
+
+func intersection(start *api.ObjectAndRelation, children ...*api.RelationTupleTreeNode) *api.RelationTupleTreeNode {
+	return setResult(api.SetOperationUserset_INTERSECTION, start, children)
+}
+
+func exclusion(start *api.ObjectAndRelation, children ...*api.RelationTupleTreeNode) *api.RelationTupleTreeNode {
+	return setResult(api.SetOperationUserset_EXCLUSION, start, children)
+}
+
+type userSet map[string]struct{}
+
+func (us userSet) Add(users ...*api.User) {
```

Comment:
Doc comments on anything exported

Reply:
It's on a non-exported type -- this isn't actually exported. I'll lowercase it.